### Task 1

In [ ]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('/content/sample_data/adult.csv')

# Initial inspection
print("First 5 rows:")
display(df.head())

print("\nDataset information:")
df.info()

# Check missing values before fixing hidden values
print("\nMissing values before replacing '?':")
print(df.isnull().sum())

# Replace hidden missing values with NaN
df.replace('?', np.nan, inplace=True)

# Check missing values after replacement
print("\nMissing values after replacing '?':")
print(df.isnull().sum())

First 5 rows:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K



Dataset information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB

Missing values before replacing '?':
age               0
workclass    

### Task 2

In [ ]:
# Count missing values per column
missing_count = df.isnull().sum()

# Calculate missing percentage per column
missing_percentage = df.isnull().mean() * 100

# Create a summary table
missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percentage
})

# Sort by missing count in descending order
missing_summary = missing_summary.sort_values(
    by='Missing Count',
    ascending=False
)

display(missing_summary)

,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486
fnlwgt,0,0.000000
education,0,0.000000
education.num,0,0.000000
age,0,0.000000
marital.status,0,0.000000
relationship,0,0.000000
sex,0,0.000000


### Task 3

In [ ]:
# Check missing values before handling them
print("Missing values before handling:")
print(df.isnull().sum())

# Handle missing categorical values
df['workclass'] = df['workclass'].fillna('Unknown')
df['occupation'] = df['occupation'].fillna('Unknown')
df['native.country'] = df['native.country'].fillna('Unknown')

# Check missing values after handling
print("\nMissing values after handling:")
print(df.isnull().sum())

# Display the number of Unknown values introduced
print("\nUnknown values after imputation:")
print("workclass:", (df['workclass'] == 'Unknown').sum())
print("occupation:", (df['occupation'] == 'Unknown').sum())
print("native.country:", (df['native.country'] == 'Unknown').sum())

Missing values before handling:
age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64

Missing values after handling:
age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64

Unknown values after imputation:
workclass: 1836
occupation: 1843
native.country: 583


### Task 4

In [ ]:
# Check for exact duplicate rows
exact_duplicates = df.duplicated().sum()

print("Number of exact duplicate rows:")
print(exact_duplicates)

# Check for duplicates while ignoring the income target/label column
columns_without_target = df.columns.drop('income')

duplicates_without_target = df.duplicated(
    subset=columns_without_target
).sum()

print("\nNumber of duplicate rows ignoring the income label:")
print(duplicates_without_target)

# Display duplicate rows ignoring the income label
print("\nDuplicate rows ignoring the income label:")
display(
    df[df.duplicated(subset=columns_without_target, keep=False)]
    .sort_values(by=list(columns_without_target))
)

Number of exact duplicate rows:
24

Number of duplicate rows ignoring the income label:
25

Duplicate rows ignoring the income label:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
19622,19,Private,97261,HS-grad,9,Never-married,Farming-fishing,Not-in-family,White,Male,0,0,40,United-States,<=50K
20507,19,Private,97261,HS-grad,9,Never-married,Farming-fishing,Not-in-family,White,Male,0,0,40,United-States,<=50K
10307,19,Private,138153,Some-college,10,Never-married,Adm-clerical,Own-child,White,Female,0,0,10,United-States,<=50K
22783,19,Private,138153,Some-college,10,Never-married,Adm-clerical,Own-child,White,Female,0,0,10,United-States,<=50K
17456,19,Private,146679,Some-college,10,Never-married,Exec-managerial,Own-child,Black,Male,0,0,30,United-States,<=50K
22934,19,Private,146679,Some-college,10,Never-married,Exec-managerial,Own-child,Black,Male,0,0,30,United-States,<=50K
7615,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
32065,19,Private,251579,Some-college,10,Never-married,Other-service,Own-child,White,Male,0,0,14,United-States,<=50K
9269,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K
14346,20,Private,107658,Some-college,10,Never-married,Tech-support,Not-in-family,White,Female,0,0,10,United-States,<=50K


### Task 5

In [ ]:
# Display unique values before cleaning
print("Unique values in education:")
print(df['education'].unique())

print("\nUnique values in marital.status:")
print(df['marital.status'].unique())

print("\nUnique values in native.country:")
print(df['native.country'].unique())

# Remove leading and trailing whitespace from categorical columns
df['education'] = df['education'].str.strip()
df['marital.status'] = df['marital.status'].str.strip()
df['native.country'] = df['native.country'].str.strip()

# Display unique values after cleaning
print("\nUnique values in education after stripping:")
print(df['education'].unique())

print("\nUnique values in marital.status after stripping:")
print(df['marital.status'].unique())

print("\nUnique values in native.country after stripping:")
print(df['native.country'].unique())

Unique values in education:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Unique values in marital.status:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Unique values in native.country:
['United-States' 'Unknown' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']

Unique values in education after stripping:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11t

### Task 6

In [ ]:
# Descriptive statistics for numeric columns
print("Descriptive Statistics for Numeric Columns:")
display(df.describe())

# Descriptive statistics for categorical columns
print("\nDescriptive Statistics for Categorical Columns:")
display(df.describe(include='object'))

# Mean and median for age
age_mean = df['age'].mean()
age_median = df['age'].median()

# Mean and median for hours-per-week
hours_mean = df['hours.per.week'].mean()
hours_median = df['hours.per.week'].median()

print("\nAge:")
print("Mean:", age_mean)
print("Median:", age_median)

print("\nHours per Week:")
print("Mean:", hours_mean)
print("Median:", hours_median)

# Compare mean and median
print("\nComparison:")
if abs(age_mean - age_median) < 1:
    print("Mean and median age are very close.")
else:
    print("Mean and median age differ meaningfully.")

if abs(hours_mean - hours_median) < 1:
    print("Mean and median hours-per-week are very close.")
else:
    print("Mean and median hours-per-week differ meaningfully.")

Descriptive Statistics for Numeric Columns:


,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000



Descriptive Statistics for Categorical Columns:


,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720



Age:
Mean: 38.58164675532078
Median: 37.0

Hours per Week:
Mean: 40.437455852092995
Median: 40.0

Comparison:
Mean and median age differ meaningfully.
Mean and median hours-per-week are very close.


### Task 7

In [ ]:
# Most common occupation
occupation_counts = df['occupation'].value_counts()

print("Occupation Frequency:")
display(occupation_counts)

print("\nMost Common Occupation:")
print(occupation_counts.idxmax())
print("Count:", occupation_counts.max())

# Percentage distribution of sex
sex_percentage = df['sex'].value_counts(normalize=True) * 100

print("\nPercentage Distribution of Sex:")
display(sex_percentage)

# Percentage distribution of income
income_percentage = df['income'].value_counts(normalize=True) * 100

print("\nPercentage Distribution of Income:")
display(income_percentage)

# Check whether income classes are balanced or imbalanced
income_difference = abs(income_percentage.iloc[0] - income_percentage.iloc[1])

print("\nIncome Class Balance:")
if income_difference < 10:
    print("The income classes are relatively balanced.")
else:
    print("The income classes are imbalanced.")

Occupation Frequency:


,count
occupation,
Prof-specialty,4140
Craft-repair,4099
Exec-managerial,4066
Adm-clerical,3770
Sales,3650
Other-service,3295
Machine-op-inspct,2002
Unknown,1843
Transport-moving,1597



Most Common Occupation:
Prof-specialty
Count: 4140

Percentage Distribution of Sex:


,proportion
sex,
Male,66.920549
Female,33.079451



Percentage Distribution of Income:


,proportion
income,
<=50K,75.919044
>50K,24.080956



Income Class Balance:
The income classes are imbalanced.


### Task 8

In [ ]:
# Check the mapping between education and education.num
education_mapping = df.groupby('education')['education.num'].unique()

print("Education to education-num mapping:")
display(education_mapping)

# Identify education levels mapped to more than one number
inconsistencies = education_mapping[
    education_mapping.apply(lambda x: len(x) > 1)
]

print("\nInconsistent education mappings:")
if len(inconsistencies) == 0:
    print("No inconsistencies found.")
else:
    display(inconsistencies)

# Final result
print("\nResult:")
if len(inconsistencies) == 0:
    print("Education and education-num are consistent.")
else:
    print("Education and education-num contain inconsistencies.")

Education to education-num mapping:


,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]



Inconsistent education mappings:
No inconsistencies found.

Result:
Education and education-num are consistent.


### Task 9

In [ ]:
import sqlite3
import pandas as pd

# Load the original CSV file
raw_df = pd.read_csv('/content/sample_data/adult.csv')

# Create a local SQLite database
conn = sqlite3.connect('/content/adult_income.db')

# Load the raw CSV data into a SQLite table
raw_df.to_sql(
    'adult_income',
    conn,
    if_exists='replace',
    index=False
)

print("Raw CSV loaded successfully into SQLite table: adult_income")

# Run SQL query
query = """
SELECT *
FROM adult_income
WHERE age > 30;
"""

# Load SQL query result into a Pandas DataFrame
sql_df = pd.read_sql_query(query, conn)

print("\nSQL Query:")
print(query)

print("First 5 rows extracted using SQL:")
display(sql_df.head())

print("\nShape of original CSV:")
print(raw_df.shape)

print("\nShape of SQL extracted DataFrame:")
print(sql_df.shape)

print("\nNumber of rows with age > 30:")
print(len(sql_df))

# Verify the extraction
print("\nSQL extraction completed successfully.")

Raw CSV loaded successfully into SQLite table: adult_income

SQL Query:

SELECT *
FROM adult_income
WHERE age > 30;

First 5 rows extracted using SQL:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K



Shape of original CSV:
(32561, 15)

Shape of SQL extracted DataFrame:
(21989, 15)

Number of rows with age > 30:
21989

SQL extraction completed successfully.


# Task 10

## Data Profiling Summary Report

### Dataset Overview

The Adult Income dataset has information about people. It includes details about where they live, what they do and how money they make. This dataset is used to figure out what things affect how money a person makes.

The Adult Income dataset has 48,842 rows and 15 columns.

### Data Quality Issues Found

When we first looked at the data we found some values that were hidden. These missing values were marked with a question mark. We replaced these question marks with a value that means "no data". Then we found out that there were missing values in the workclass, occupation and native country columns.

* The workclass column has 1,836 missing values

* The occupation column has 1,843 missing values

* The native country column has 583 missing values

We also checked the Adult Income dataset for rows and made sure the categories were correct. We removed spaces from the beginning and end of the education marital status and native country columns.

### Key Observations

We calculated some statistics for the Adult Income dataset. We looked at the middle values of the age and hours worked per week columns.

We also did a frequency analysis to see what the common job was and how many men and women there were in the Adult Income dataset. We looked at how money people made and found out that the income classes were not balanced. This means that there were people who made less than 50,000 dollars than people who made more than 50,000 dollars.

We checked to make sure the education and education number columns were consistent with each other.

### Cleaning Decisions Made

We replaced the hidden question marks with a no data" value so we could find the missing values. We replaced the missing values in the workclass, occupation and native country columns with "Unknown" of removing those rows.

We removed spaces from the education, marital status and native country columns. We checked for rows but we did not remove any rows that had the same information but different income.

Finally we used a database to extract the data and then loaded it back, into a program called Pandas for analysis. The Adult Income dataset was used for this analysis.